DataSourceAzure: fix bug where reported ready marker file is being written before report ready in preprovisioning

johnsonshi opened the pull request:
## Proposed Commit Message

DataSourceAzure: Write reported ready marker file after report ready in preprovisioning

DataSourceAzure previously writes the preprovisioning
reported ready marker file before it goes through the
report ready workflow. On certain VM instances, the
marker file is successfully written but then reporting
ready fails.

Upon rare VM reboots by the platform, cloud-init sees
that the report ready marker file already exists.
The existence of this marker file tells cloud-init
not to report ready again (because it mistakenly
assumes that it already reported ready in
preprovisioning).

In this scenario, cloud-init instead erroneously
takes the reprovisioning workflow instead of
reporting ready again.

## Additional Context
Additional context is already in the proposed commit message.

## Test Steps
* Create an Azure VM from the latest Ubuntu 18.04 LTS image.
* Apply this PR's patch.
* Capture the Azure VM into a custom image.
* Successful deployment from this captured custom image.

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes
that apply. -->
 - [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/topics/hacking.html)
 - [x] I have updated or added any unit tests accordingly
 - [x] I have updated or added any documentation accordingly

johnsonshi commented:
@anhvoms for awareness

OddBloke commented on cloudinit/sources/DataSourceAzure.py:
```diff
@@ -693,12 +693,22 @@ def exc_cb(msg, exception):
+                    if not report_ready_succeeded:
+                        msg = ('Failed reporting ready while in '
+                               'the preprovisioning pool.')
+                        report_diagnostic_event(msg)
```
Are we intentionally not passing a logger in these `report_diagnostic_event` calls?

johnsonshi replied:
Awaiting the merging of #563 before I add that here.

johnsonshi replied:
Merged the code from the branch of #563 to this PR. It now has `logger_func`.

OddBloke reviewed (commented):
This looks good, thanks! A few style/logging Qs inline.

johnsonshi commented on tests/unittests/test_datasource/test_azure.py:
```diff
@@ -2074,9 +2086,47 @@ def test_poll_imds_report_ready_false(self,
+        self.assertFalse(os.path.exists(report_file))
+
+
+@mock.patch(MOCKPATH + 'DataSourceAzure._report_ready')
```
Ahhh I didn't know this was possible. Thanks!

johnsonshi commented:
Hey @OddBloke, thanks for the review on this PR. I've addressed your comments (regarding the mocks on the UTs + regarding the `report_diagnostic_event`'s `logger_func` args for the new diagnostic messages).

The last time you reviewed this PR, this PR was small. This PR may deceivingly seem to have blown up in size since then, but it's only because I've merged in the code from #563 to this branch. Once #563 is merged, it will go back to the previous small size.

johnsonshi commented on cloudinit/sources/DataSourceAzure.py:
```diff
@@ -773,14 +784,20 @@ def exc_cb(msg, exception):
         try:
             get_metadata_from_fabric(None, lease['unknown-245'])
+            return True
         except Exception as e:
```
To address any potential concerns: This works, since `get_metadata_from_fabric` always raises an exception within the implemented child functions of `get_metadata_from_fabric`.

[code](https://github.com/canonical/cloud-init/blob/3b05b1a6c58dfc7533a16f795405bda0e53aa9d8/cloudinit/sources/helpers/azure.py#L914-L922)
[code](https://github.com/canonical/cloud-init/blob/72f1d8df96a9e14444f96363c9b08845bbfe265b/tests/unittests/test_datasource/test_azure_helper.py#L744-L751)
[code](https://github.com/canonical/cloud-init/blob/0d5ac8f035c78971f7857a67bf9ce637a6798c77/tests/unittests/test_datasource/test_azure_helper.py#L709-L727)

OddBloke commented on tests/unittests/test_datasource/test_azure.py:
```diff
@@ -2080,11 +2093,10 @@ def fake_timeout_once(**kwargs):
-        """The poll_imds should not call reporting ready
-           when flag is false"""
+    def test_does_not_poll_imds_report_ready_when_marker_file_exists(
+            self, m_report_ready, m_request, m_media_switch, m_dhcp, m_net):
```
Thanks for these renames!

OddBloke reviewed (approved):
Thanks Johnson, this looks good to me. I particularly appreciate the clear and complete testing code, it made it much easier to determine whether or not you had modified behaviour intentionally!
